src/ngAnimate/animateQueue.js

```diff
@@ -0,0 +1,558 @@
+          var details = activeAnimationsLookup.get(node);
+          if (details && details.counter === counter) {
+            clearElementAnimationState(element);
+          }
+          notifyProgress(runner, event, 'close', {});
+        });
+
+        // this will update the runner's flow-control events based on
+        // the `realRunner` object.
+        runner.setHost(realRunner);
+        notifyProgress(runner, event, 'start', {});
+      });
+
+      return runner;
+
+      function notifyProgress(runner, event, phase, data) {
+        triggerCallback(event, element, phase, data);
+        runner.progress(event, phase, data);
+      }
+
+      function close(reject) { // jshint ignore:line
+        options.$applyClasses();
+        options.$applyStyles();
+        options.$domOperation();
+        runner.complete(!reject);
+      }
+    }
+
+    function closeChildAnimations(element) {
+      var node = element[0];
+      var children = node.querySelectorAll('[' + NG_ANIMATE_ATTR_NAME + ']');
+      forEach(children, function(child) {
+        var state = parseInt(child.getAttribute(NG_ANIMATE_ATTR_NAME));
+        var details = activeAnimationsLookup.get(child);
+        switch (state) {
+          case RUNNING_STATE:
+            details.runner.end();
+            /* falls through */
+          case PRE_DIGEST_STATE:
+            if (details) {
+              activeAnimationsLookup.remove(child);
+            }
+            break;
+        }
+      });
+    }
+
+    function clearElementAnimationState(element) {
+      element = element.length ? element[0] : element;
+      element.removeAttribute(NG_ANIMATE_ATTR_NAME);
+      activeAnimationsLookup.remove(element);
+    }
+
+    function isMatchingElement(a,b) {
+      a = a.length ? a[0] : a;
+      b = b.length ? b[0] : b;
+      return a === b;
+    }
+
+    function closeParentClassBasedAnimations(startingElement) {
+      var parentNode = startingElement[0];
+      do {
+        if (!parentNode || parentNode.nodeType !== ELEMENT_NODE) break;
+
+        var details = activeAnimationsLookup.get(parentNode);
+        if (details) {
+          examineParentAnimation(parentNode, details);
+        }
+
+        parentNode = parentNode.parentNode;
+      } while (true);
+
+      // since animations are detected from CSS classes, we need to flush all parent
+      // class-based animations so that the parent classes are all present for child
+      // animations to properly function (otherwise any CSS selectors may not work)
+      function examineParentAnimation(node, animationDetails) {
+        // enter/leave/move always have priority
+        if (animationDetails.structural) return;
+
+        if (animationDetails.state === RUNNING_STATE) {
+          animationDetails.runner.end();
+        }
+        clearElementAnimationState(node);
+      }
+    }
+
+    function areAnimationsAllowed(element, parent, event) {
+      var bodyElement = jqLite($document[0].body);
```

Comment:
Couldn't this be moved outside of the function (for performance reasons) ?
What are the implications ? Can the body really change ?